[IMP] runbot: add without-demo as field

Williambraecky opened the pull request:
Adds the without-demo flag as a field on config steps.

Williambraecky committed:
[IMP] runbot: add without-demo as field

Adds the without-demo flag as a field on config steps.

Williambraecky commented on runbot/models/build_config.py:
```diff
@@ -431,6 +432,11 @@ def _run_install_odoo(self, build):
             build._local_pg_createdb(db_name)
         cmd += ['-d', db_name]
+        # Demo data behavior changed in 18.1 -> demo data became opt-in instead of opt-out
+        if build.version_id.number >= '18.01' and self.with_demo:
```
This will not work obviously with the PR but I have no idea how we can make it work "temporarily"

Xavier-Do commented on runbot/models/build_config.py:
```diff
@@ -160,6 +160,7 @@ class ConfigStep(models.Model):
     extra_params = fields.Char('Extra cmd args', tracking=True)
     additionnal_env = fields.Char('Extra env', help='Example: foo="bar";bar="foo". Cannot contains \' ', tracking=True)
     enable_log_db = fields.Boolean("Enable log db", default=True)
+    with_demo = fields.Boolean("Install demo data", default=True, tracking=True)
```
I'm not sure that having a default True is correct since we may want to have the default behaviour most of the time. (single module, testing, install of base database, at install, ....)

We may want to be able to force without in previous version, in this case the flag --without is already added 
We may want to force with demo on future version, but the only use case would be to have a database for the run (and maybe a nightly) and this will be a contionnal step. (have a database with demo for the running build) 

But adding it on the config could be a nice idea, I think that in this case we should also add support for `build.params_id.config_params.get('no_demo')`. But the field may be a selection field (default, no_demo, demo) so that the odoo testing can run without demo. Upgrade template will force "no demo" and master default config could install a template with demo on the side for the running build.

Xavier-Do replied:
Another solution could be to have two boolean fields
without-demo and with demo, both of them being False by default.

Williambraecky replied:
Thanks for the insight 😄 

I like the selection field as it provides the most flexibility so I implemented it and tested all different ways.
Let me know what you think :)

Xavier-Do replied:
The selection field doesn't allow an incorrect state where the two are True but it could have a 4th effect by design, do the opposite (no demo if demo is default, and vice versa)

Williambraecky replied:
Do you want me to add that 4th state?

Xavier-Do replied:
Not sure. Would be confusing. We can talk about that wednesday

Xavier-Do commented on runbot/models/build_config.py:
```diff
@@ -768,6 +774,11 @@ def _run_test_upgrade(self, build):
         upgrade_paths = list(build._get_upgrade_path())
         if upgrade_paths:
             migrate_cmd += ['--upgrade-path', ','.join(upgrade_paths)]
+        # Demo data behavior changed in 18.1 -> demo data became opt-in instead of opt-out
+        if build.version_id.number >= '18.01' and self.with_demo:
+            migrate_cmd.append('--with-demo')
+        elif build.version_id.number < '18.01' and not self.with_demo:
+            migrate_cmd.append('--without-demo=true')
```
```suggestion
        # Demo data behavior changed in 18.1 -> demo data became opt-in instead of opt-out
        available_options = build._parse_config()
        if self.with_demo and '--with-demo' in available_options:
            migrate_cmd.append('--with-demo')
        elif not self.with_demo and '--with-demo' not in available_options:
            migrate_cmd.append('--without-demo=true')
```

Xavier-Do commented on runbot/models/build_config.py:
```diff
@@ -431,6 +432,11 @@ def _run_install_odoo(self, build):
             build._local_pg_createdb(db_name)
         cmd += ['-d', db_name]
+        # Demo data behavior changed in 18.1 -> demo data became opt-in instead of opt-out
+        if build.version_id.number >= '18.01' and self.with_demo:
+            cmd.append('--with-demo')
+        elif build.version_id.number < '18.01' and not self.with_demo:
+            cmd.append('--without-demo=true')
```
```suggestion
        available_options = build._parse_config()
        if self.with_demo and '--with-demo' in available_options:
            migrate_cmd.append('--with-demo')
        elif not self.with_demo and --with-demo' not in available_options:
            migrate_cmd.append('--without-demo=true')
```

Williambraecky committed:
[IMP] runbot: add config_data to triggers

Adds support for config_data directly on the triggers.
This will be used to share configs between triggers with different
parameters.

Xavier-Do commented on runbot/models/build_config.py:
```diff
@@ -431,13 +435,24 @@ def _run_install_odoo(self, build):
         cmd += ['-d', db_name]
+
+        # Demo data behavior changed in 18.1 -> demo data became opt-in instead of opt-out
+        available_options = build._parse_config()
+        build_default_demo_behavior = 'without_demo' if '--with-demo' in available_options else 'with_demo'
+        demo_mode = build.params_id.config_data.get('demo_mode', self.demo_mode)
+        wanted_demo_behavior = build_default_demo_behavior if demo_mode == 'default' else demo_mode
+        if wanted_demo_behavior != build_default_demo_behavior:
+            if wanted_demo_behavior == 'with_demo':
+                cmd.append('--with-demo')
+            else:
+                cmd.append('--without-demo=true')
+
```
```suggestion
        available_options = build._parse_config()
        demo_with_is_default = '--with-demo' not in available_options:
        demo_mode = build.params_id.config_data.get('demo_mode', self.demo_mode)
        if wanted_demo_behavior != build_default_demo_behavior:
            if demo_mode == 'with_demo' and not demo_with_is_default:
                cmd.append('--with-demo')
            elif demo_mode == 'without_demo'  and demo_with_is_default
                cmd.append('--without-demo=true')
```

Not sure, maybe?

Xavier-Do commented on runbot/models/build_config.py:
```diff
@@ -160,6 +160,10 @@ class ConfigStep(models.Model):
     enable_log_db = fields.Boolean("Enable log db", default=True)
+    demo_mode = fields.Selection(
+        [('default', 'Default'), ('without_demo', 'Without Demo'), ('with_demo', 'With Demo')],
+        "Install demo data", default='default', tracking=True
```
`required=True`
